### ISSUE TITLE
Staff users can edit their own permissions

### ISSUE THREAD

**Issue by tomkins** [django-guardian/django-guardian#396]
Related to #305 and cdb251c52cc24b823d00a05da04d073568d690b0 - maybe.
- Superuser gives access to staff user, with edit object permission - "Can change object", and no other permissions
- Staff user goes to the object, clicks permissions, and gives themselves other permissions for that object (eg. delete)

Fortunately the Django admin doesn't check for object level permissions when it comes to object deletion - it requires the model delete permission.

However this staff user can add/edit/remove permissions for any other user for this object, and my site code is checking object level permissions - this staff user could maliciously add whatever permissions they want.

Should staff users be able to edit permissions on an object?

**Comment by brianmay** [comment]
The assumption has always been that if you grant people access to the admin site, they have access to everything. As such the answer to your question is Yes, staff should should be able to edit permissions. Especially if they have the can change permission for the object class. As you have noted the admin class doesn't look at object level permissions.

It is possible for your specific application this decision is not appropriate. In which case you need to decide if using the admin interface really is the best solution and/or if you can override appropriate parts to get the behavior you desire.

Does this help?

**Comment by brianmay** [comment]
No response. Closing. Can reopen if required.

**Comment by hmpf** [comment]
I have looked into this. For others who are curious:

There are model permissions and object permissions. In the admin, in order for the object permission to work as expected, the same user *also* need a model permission. What I mean with expected is: When you log in to the admin, in order to see any apps in the list of apps and models, you need at least one of "add", "change" or "delete" permissions **on the model itself**. In order for the "Delete <model>" action to work, you also need a "delete_<model>" model-permission or you'll end up with a PermissionDenied when you hit Go.

What the object permissions allow, in the admin, is direct access to the complete url. Object access, not queryset access. If you access an object by direct url in the admin and have an object delete permission, there will be a Delete-button but as of Django 1.11.x and django-guardian 1.4.9 that button will not work (TypeError: collect() got multiple values for argument 'using').

To sum up: 

1. The django admin doesn't fully support object permissions, as even the action "delete_selected" would have to be rewritten. It doesn't take the queryset into account before it checks for permission.
2. Admin-integration is poorly documented. The queryset is altered by ownership (a foreign key to User or Group). This is not documented in the admin integration docs. (I would need a ManyToManyField for at least the user, anyway.)

**Comment by ad-m** [comment]
@hmpf , could you provide pull requests to improve django-admin integration based on your analysis?

**Comment by thclark** [comment]
Closing in favour of #756
